Fix MANIFEST.MF parsing

Kationor opened the pull request:
Related to #3122 (CRLF-encoded manifests cause a trailing CR in the version).

~~Not a great solution, but the simplest I could think of. (others include using QStrings split() with a regex, simply special-casing the carriage return, and normalizing EOLs beforehand)~~

~~This definitely does not comply with the manifest specification, but it's not like the previous implementation did either.~~

TheKodeToad reviewed (changes requested):
I would prefer changing out the split char to the regex `\r?\n`

Kationor commented:
> I would prefer changing out the split char to the regex `\r?\n`

I've gone with `\r\n|\n|\r` because that matches [the spec](https://docs.oracle.com/en/java/javase/21/docs/specs/jar/jar.html#name-value-pairs-and-sections).

While I was reading the spec, I noticed the comment on property names being case insensitive, so I decided to implement that as well. This should mean that we can now parse *most* valid manifests with the exception of manifests using multi-line values.

Kationor commented:
Quick before-and-after on a few test jars (using the modlist export):

Before:
```
CR-Encoded [NONE]
CRLF-Encoded [1.0.0
]
CRLF-Encoded + Lowercase [NONE]
LF-Encoded [1.0.0]
```

After:
```
CR-Encoded [1.0.0]
CRLF-Encoded [1.0.0]
CRLF-Encoded + Lowercase [1.0.0]
LF-Encoded [1.0.0]
```

TheKodeToad reviewed (approved):
Unsure which any mods only have the version in MANIFEST.MF... but looks good!

Kationor commented:
> Unsure which any mods only have the version in MANIFEST.MF... but looks good!

Excerpt from the `mods.toml` of a mod I've been using to test:
```toml
# A URL to refer people to when problems occur with this mod
#issueTrackerURL="https://change.me.to.your.issue.tracker.example.invalid/" #optional
# A list of mods - how many allowed here is determined by the individual mod loader
[[mods]] #mandatory
modId = "adaptive_performance_tweaks_spawn"
version = "${file.jarVersion}"
displayURL = "https://www.curseforge.com/minecraft/mc-mods/adaptive-performance-tweaks-spawn/"
displayName = "APTweaks: Spawn"
displayTest = "IGNORE_SERVER_VERSION"
logoFile = "logo.png"
credits = ""
authors = "Markus Bordihn"
description = '''Adaptive Performance Tweaks Spawn automatically optimize spawn events and spawner.''
```

Trial97 commented on launcher/minecraft/mod/tasks/LocalModParseTask.cpp:
```diff
@@ -487,11 +488,11 @@ bool processZIP(Mod& mod, [[maybe_unused]] ProcessingLevel level)
 
                 // quick and dirty line-by-line parser
-                auto manifestLines = file.readAll().split('\n');
+                auto manifestLines = QString(file.readAll()).split(QRegularExpression("\r\n|\n|\r"));
```
I would suggest that you declare this regular expression as a static global variable

Kationor committed:
Improve MANIFEST.MF parsing

Previously, we would only properly parse LF-encoded manifests, and even
those only if they used the recommended casing.

This commit allows the parser to recognise CR and CRLF newlines, and
also makes the name comparison case insensitive to align with the
specification. (Though not completely: we still don't support multiline
values)

Signed-off-by: Kationor <n96211028@gmail.com>

Trial97 commented:
@Kationor does this also fix the mentioned issue? Or do I need an extra trim on modlist export?

Kationor commented:
> @Kationor does this also fix the mentioned issue? Or do I need an extra trim on modlist export?

#3122 was a symptom of Prism not handling the newlines correctly, so that's fixed by this PR. Trimming leading/trailing whitespace from the version might still make sense, but that's a separate thing.

TheKodeToad commented:
> > Unsure which any mods only have the version in MANIFEST.MF... but looks good!
> 
> Excerpt from the `mods.toml` of a mod I've been using to test:
> 
> ```toml
> # A URL to refer people to when problems occur with this mod
> #issueTrackerURL="https://change.me.to.your.issue.tracker.example.invalid/" #optional
> # A list of mods - how many allowed here is determined by the individual mod loader
> [[mods]] #mandatory
> modId = "adaptive_performance_tweaks_spawn"
> version = "${file.jarVersion}"
> displayURL = "https://www.curseforge.com/minecraft/mc-mods/adaptive-performance-tweaks-spawn/"
> displayName = "APTweaks: Spawn"
> displayTest = "IGNORE_SERVER_VERSION"
> logoFile = "logo.png"
> credits = ""
> authors = "Markus Bordihn"
> description = '''Adaptive Performance Tweaks Spawn automatically optimize spawn events and spawner.''
> ```

*why any mods

Yeaah the whole `${file.jarVersion}` looks like something which would only happen if variable replacement failed or something but it might be some handling fml does

evan-goode commented:
Nice. If there's a need for an even more robust parser in the future, I ported the OpenJDK Manifest.mf parser here for Fjord Launcher, to identify certain Java agent JARs: https://github.com/unmojang/FjordLauncher/blob/develop/launcher/Manifest.cpp [edited: please do not link a piracy fork]. (After trying and failing to use a C++ header-only parser generator and implementing it by the spec).

TheKodeToad commented:
Personally I'm all for implementing parsing correctly instead of doing hacks with strings. You can rest assured that there will be no more issues with edge cases.

We're also using INI parsing for Java properties files, although I don't think these are nearly as common and I don't know of any issues which have arised from this (besides the fact that the INI parser does require actual files - a string is not enough - so a tmp file is used).

I don't know the opinion of other maintainers though. I can say though unless there are other issues open about parsing problems I don't think there's really any need, for now.

P.S. you seemed to link a fork of your fork. since there was no actual reason for this I will just assume it was a mistake :P

evan-goode commented:
Whoops, thanks for fixing the link. Picked the first Google result and didn't look... 

Yeah, definitely not an urgent thing. Just wanted to make it known and maybe I'll open a PR if I have some spare time.
